libs/ledger/include/ledger/dag/dag.hpp

```diff
@@ -0,0 +1,204 @@
+
+  // Revert to a previous/forward epoch
+  bool RevertToEpoch(uint64_t) override;
+
+  uint64_t CurrentEpoch() const override;
+
+  bool HasEpoch(EpochHash const &hash) override;
+
+  // Make sure that the dag has all nodes for a certain epoch
+  bool SatisfyEpoch(DAGEpoch &) override;
+
+  std::vector<DAGNode> GetLatest(bool previous_epoch_only = false) override;
+
+  ///////////////////////////////////////
+  // Fns used for syncing
+
+  // Recently added DAG nodes by this miner (not seen by network yet)
+  std::vector<DAGNode> GetRecentlyAdded() override;
+
+  // DAG node hashes this miner knows should exist
+  MissingNodeHashes GetRecentlyMissing() override;
+
+  bool GetDAGNode(ConstByteArray const &hash, DAGNode &node) override;
+  bool GetWork(ConstByteArray const &hash, Work &work) override;
+
+  // TXs and epochs will be added here when they are broadcast in
+  bool AddDAGNode(DAGNode node) override;
+
+private:
+  // Long term storage
+  uint64_t             most_recent_epoch_ = 0;
+  DAGEpoch             previous_epoch_;   // Most recent epoch, not in deque for convenience
+  std::deque<DAGEpoch> previous_epochs_;  // N - 1 still relevant epochs
+  EpochStackStore      epochs_;  // Past less-relevant epochs as a stack (key = index, value = hash)
+  EpochStore all_stored_epochs_;  // All epochs, including from non-winning forks (key = epoch hash,
+                                  // val = epoch)
+  DAGNodeStore finalised_dnodes_;  // Once an epoch arrives, all dag nodes inbetween go here
+  DAGEpoch     temp_recently_created_epoch_;  // Most recent epoch, not in deque for convenience
+
+  // volatile state
+  std::unordered_map<DAGTipID, DAGTipPtr> all_tips_;  // All tips are here
+  std::unordered_map<NodeHash, DAGTipPtr>
+      tips_;  // lookup tips of the dag pointing at a certain node hash
+  std::unordered_map<NodeHash, DAGNodePtr>
+      node_pool_;  // dag nodes that are not finalised but are still valid
+  std::unordered_map<NodeHash, DAGNodePtr>
+      loose_nodes_;  // nodes that are missing one or more references (waiting on NodeHash)
+  std::unordered_map<NodeHash, std::vector<DAGNodePtr>>
+      loose_nodes_lookup_;  // nodes that are missing one or more references (waiting on NodeHash)
+
+  // std::unordered_map<NodeHash, uint64_t>                loose_nodes_ttl_;   // TODO(HUT): loose
+  // nodes management scheme
+
+  // Used for sync purposes
+  std::vector<DAGNode> recently_added_;  // nodes that have been recently added
+  std::set<NodeHash>   missing_;         // node hashes that we know are missing
+
+  // Internal functions don't need locking and can recursively call themselves etc.
+  bool       PushInternal(DAGNodePtr node);
+  bool       AlreadySeenInternal(DAGNodePtr node);    // const
+  bool       TooOldInternal(uint64_t);                // const
+  bool       IsLooseInternal(DAGNodePtr node);        // const
+  void       SetReferencesInternal(DAGNodePtr node);  // const
```

Comment:
`const`?

Reply:
the object store doesn't have const methods making accesses to it not const. It's on the radar

Reply:
@n-hutton can we remove the `//const` and raise a tracking issue for this

Reply:
yes